### PR TITLE
#1269 Fix OOM when using sparkey side input

### PR DESCRIPTION
**Pull request by jto**
Fix for #1269.
Not sure if I should keep the it test. Running it takes forever and it's only useful if you run it, monitor the heap usage and force GC at some point to see if it's consuming too much memory....

### CODE REVIEW

File: scio-core/src/main/scala/com/spotify/scio/values/SideInput.scala [hunk 1]
```diff
@@ -32,10 +32,14 @@ trait SideInput[T] extends Serializable {
...
 
+  // Use this attribute in implementations of SideInput to force caching
+  // even on GlobalWindows. (Used to fix #1269)
+  protected def forceCache = false
+
   private[values] def get[I, O](context: DoFn[I, O]#ProcessContext): T
 
   def getCache[I, O](context: DoFn[I, O]#ProcessContext, window: BoundedWindow): T = {
-    if (cache == null || window == GlobalWindow.INSTANCE || this.window != window) {
+    if (cache == null || (!forceCache && window == GlobalWindow.INSTANCE) || this.window != window) {
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/SideInput.scala]: Failed stylecheck
`/home/circleci/scio/scio-core/src/main/scala/com/spotify/scio/values/SideInput.scala:42: File line length exceeds 100 characters`

Comment (jto) [reply scio-core/src/main/scala/com/spotify/scio/values/SideInput.scala]: fixed

**Comment by regadas** [comment]
@jto I was just wondering if renaming `forceCache` to `forceCacheOnGlobalWindow` would be a better name?

File: scio-extra/src/it/scala/com/spotify/scio/extra/sparkey/SparkeyIT.scala [hunk 2]
```diff
@@ -50,8 +50,35 @@ class SparkeyIT extends PipelineSpec {
...
+          }.toSCollection
+      } finally {
+        FileSystems.delete(Seq(resourceId).asJava)
+      }
+    }
+  }
+
   it should "throw exception when Sparkey file exists" in {
     runWithContext { sc =>
+      import com.spotify.scio.extra.sparkey._
```
Comment (nevillelyh) [inline scio-extra/src/it/scala/com/spotify/scio/extra/sparkey/SparkeyIT.scala]: minor: redundant import?

File: scio-core/src/main/scala/com/spotify/scio/values/SideInput.scala [hunk 3]
```diff
@@ -32,10 +32,18 @@ trait SideInput[T] extends Serializable {
...
+  protected def forceCache = false
+
   private[values] def get[I, O](context: DoFn[I, O]#ProcessContext): T
 
   def getCache[I, O](context: DoFn[I, O]#ProcessContext, window: BoundedWindow): T = {
-    if (cache == null || window == GlobalWindow.INSTANCE || this.window != window) {
+    if (
+        cache == null ||
+        (!forceCache && window == GlobalWindow.INSTANCE) ||
+        this.window != window
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/SideInput.scala]: minor: group `cache == null || this.window != window` so it's easier to follow?

**Pull request merged**